### PR TITLE
Add sidecar service

### PR DESCRIPTION
**Pull request by raintygao**
### Description

Support a new layout service called sidecar in dashboard, so that external plugin can use this service as its container. This flyout supports three modes and resizable.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5620

## Screenshot

image
image
image

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### Test steps
1. Open sidecar: You could open sidecar in plugin or other application through `core.overlays.sidecar().open`, in this method you need to pass MountPoint and sidecar config, like
 `{
          className: 'chatbot-sidecar',
          config: {
            dockedMode: 'right',
            paddingSize: 460
    }`.
After that you could see a sidecar container is mounted.
2. You could drag border button to view resizable effect.
3. You could change docked Mode or docked padding size through calling `setSidecarConfig` with new config, like
`
  {
      dockedMode: 'takeover',
      paddingSize: 500
   }
`
4. You could call `hide` and `show` to implement sidecar hide and show effect.

Here is a test env. http://k8s-opensear-opensear-b09791709c-480119112.us-west-1.elb.amazonaws.com/app/home#/,  username: `admin`, password: `5fIavN2oFmWiy5E`

### CODE REVIEW

**Comment by kavilla** [comment]
Pretty cool.

Will check out the implementation. In the meantime I see there are some snapshots are not updated for tests. Also wanted to check. Is it intentional that the sidecar pushes the elements over when it is mounted?

**Comment by raintygao** [comment]
Thank you Kavilla. 

>  In the meantime I see there are some snapshots are not updated for tests. Also wanted to check. 

I will do some updates for tests.

> Is it intentional that the sidecar pushes the elements over when it is mounted?

Yes. When plugin calls  `open` function, the sidecar will be mounted with the `paddingSize` passed to push elements over.

**Comment by raintygao** [comment]
Thank you.@kavilla

>  In the meantime I see there are some snapshots are not updated for tests. Also wanted to check. 

I will do some updates to snapshots.

> Is it intentional that the sidecar pushes the elements over when it is mounted?

Yes. When plugin calls  `open` function, the sidecar will be mounted with the `paddingSize` in passed parameter to push elements over.

File: src/plugins/opensearch_dashboards_react/public/overlays/create_react_overlays.tsx
Comment (ashwin-pc) [inline src/plugins/opensearch_dashboards_react/public/overlays/create_react_overlays.tsx:1]: Can you also update the Readme for this plugin to include this feature so that it is visible easily to developers?

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (ashwin-pc) [inline src/core/public/overlays/sidecar/sidecar_service.tsx:1]: Can you add some tests for this service similar to the other two overlays?

File: src/core/public/overlays/sidecar/rsizable_button.scss
Comment (ashwin-pc) [inline src/core/public/overlays/sidecar/rsizable_button.scss:1]: Looks like you have a typo in the file name

File: src/core/public/overlays/sidecar/resizable_button.tsx
Comment (ashwin-pc) [inline src/core/public/overlays/sidecar/resizable_button.tsx:1]: Hmm, this is a custom component. Why arent we using a component from OUI for this? There is a need to make Flyouts resizeable in discover too. Instead of building a custom resizeable flyout here, we should update the flyout component in OUI to do the same. Even adding the ability to have a flyout that opens from the bottom is needed, so making a change there is the right thing to do.

**Review by ashwin-pc — COMMENTED** [review]
Nice job extending the existing overlays service to expose this feature although i wonder if this is the right approach? This just looks like a resizable flyout to me which can also open from the bottom. If thats what the sidecar is, why cant we just update the Flyout component on OUI? This is a feature that we also need for Discover and potentially in a few other places around the app. By enhancing the flyout in OUI, we avoid needing to declare a new service in core entirely and building a custom resizeable component as you have done here. 

cc: @kgcreative @shanilpa 

As for the PR itself, I havent reviewed the code yet but had a few high level comments based on the changes:
1. Can you add tests to the feature? The codecov has highlighted a few good places to do so.
2. Can you update the readme on how to use this feature. The other overlay services have sections in the readme on how to use them. This will really make this feature a lot easier to use

**Comment by ashwin-pc** [comment]
@kgcreative but cant the two be the same OUI component? They share a lot of features. Dont you think we can combine the two into one interface that is flexible as either? Here are some similarities.

| Flyout | Sidecar |
|-------| ---------|
| Lives on document root | Lives on document root |
| Opens from the side | opens from sides and bottm |
| Push and overlay | Only push |
| wants resize | Supports resize |
| ... | ... |

And even of they cant be the same component, the raw component should be bult in OUI and consumed hee like the flyout and Modal. Neither are built in OSD.

**Comment by raintygao** [comment]
@ashwin-pc 
Although sidecar and flyout have some similarities, they are actually totally different. And these two components should be able to exist at the same time without affecting each other.
The push functionality of flyout and sidecar are two completely different things. Push of flyout is implmented by adding `padding` attributes to body dom, but this can only affect app-wrapper on the OSD and cannot affect the chrome header after testing.  However,  sidecar needs to push the entire OSD. Due to the structure of the OSD dom, the component must affect on header and app-wrapper separately to implement the push of the entire OSD, which is unreasonable for an external common component. 

I think it's reasonable for OUI to support a single resizable container like [some others](https://eui.elastic.co/#/layout/resizable-container#custom-resize-logic), but if it involves interacting with entire OSD, i think this should be handled internally by the OSD, this should be more of a specific component inside the OSD.  It may be confused and more difficult to put it as a universal component in common UI library.

File: src/plugins/opensearch_dashboards_react/public/overlays/create_react_overlays.tsx
Comment (raintygao) [reply src/plugins/opensearch_dashboards_react/public/overlays/create_react_overlays.tsx:1]: Sure.

File: src/core/public/overlays/sidecar/resizable_button.tsx
Comment (raintygao) [reply src/core/public/overlays/sidecar/resizable_button.tsx:1]: 1. When I implemented this feature, OUI didn't export this component. I'm not sure whether the date of OUI support is early or later than the target version. so I choose to implement by myself.  I think this component can be removed and replaced with a component imported from OUI after OUI supports it.
2. I think update a resizable container in OUI is reasonable, flyout is only one kind of container, i'm not sure whether we need to support resizable for a specific kind container in OUI.
3. Flyout currently only supports expansion from right side. I think expanding from the bottom is more like modal rather than adding more direction to flyout.

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (raintygao) [reply src/core/public/overlays/sidecar/sidecar_service.tsx:1]: Sorry the test wasn't updated together. Since this PR needs to be merged before 2.13, so I planed to create PR for code review first. I will add tests later as soon as possible.

File: src/core/public/overlays/sidecar/resizable_button.tsx
Comment (kgcreative) [reply src/core/public/overlays/sidecar/resizable_button.tsx:1]: For what it's worth, I do agree this should be an OUI component. There are a few nuances between flyout and sidecar, but they are mostly visual.

**Comment by ashwin-pc** [comment]
@raintygao Okay i see the difference. I no longer think that the two need to be the same component, but I do still think that the component needs to come from OUI. So my hesitation with the approach is that we are defining a lot of standard UI interaction patterns in OSD for resizing that should ideally be taken care of by OUI. 

As for the mount point, I think the way the flyout does it is not ideal but doesnt mean we have to do it the same way. You can always pass a ref to a react component so that it can mount correctly. In this case the app container can be provided to an OUI Sidecar component and it can mount appropriately. All its states and combinations are also easier to test and validate while also abstracting away the accessibility and ui interaction code from the OSD codebase. not sure what your performance concerns are but lets chat about it.

> I think component in generic UI library component should not support these custom functions.

Yes. I expects similar separation of concerns similar to how the flyout or modal behaves. The UI Library is only responsible for the UI and resize properties. The state management and what to do once actions are clicked should all be handled by the sidecar service that you have here. It also makes testing these a lot easier since UI tests are quite different from the integ tests that the service needs.

**Comment by raintygao** [comment]
@ashwin-pc 
Thanks. I agree with you that some standard UI components should be imported from OUI, like resizable button, resizable flyouts. Once OUI support these components we can use them to replace raw implementation in OSD. 

I think sidecar service should be implemented in OSD instead of OUI because it is not a widely used service and currently only used by assistant plugin. Many important parts are handled in this service, such as how chrome header and app-wrapper respond respectively when the container is dragged to implement the overall push effect, some state management, some real-time configuration switching, display and hiding and many customization functions, etc.

**Comment by ashwin-pc** [comment]
+1 Can you work on adding these to OUI? Lemme know if you need any help contributing that change :)

**Comment by raintygao** [comment]
I don't know if there is a team support new features in OUI, if not, I can work on supporting this. Since I have ever not been working on OUI codebase, it may take some time. My team is rushing on releasing workspace feature on 2.13 so I may not have extra time to support this until 2.14 at the earliest.
Sidecar feature is a high property feature which was implemented in 2.12 and planned to be released in 2.13. 
I think what we discussed above is not about design or performance issues, but more like standards for code and component reuse. I agree with you that it should be updated, but its priority does not affect the release and feature delivery, especially I don't have the extra time currently, right?
So my suggestion is that we review and merge current implementation and try to support new UI components in OUI from the next version. When completed, we will replace raw implementation in OSD. I think this is a balanced solution.

**Comment by ashwin-pc** [comment]
Speaking of testing, I see that you have not included any details in the testing section of the PR Description. Can you add details in there on how to test this change?

**Comment by raintygao** [comment]
Sure, i will update test cases and test sections soon. Sorry for the missing test part,  i'm trying hard to make time for this. 

In terms of code structure, this component is indeed defined within the service, but it is also the similar to other common UI components. I will add tests for the UI components.

I really agree with you  to insist on higher standards. In this case I think it is not a flaw, it is more like a common processes: Found a widly used component containing UI patterns can be reused -> Add in the component library -> Replace the native implementation with  component imported from library. I prefer to divide it into enhancement or refactor. Of course, I have promised that I will follow up and complete it in OUI after merge.

**Comment by raintygao** [comment]
Hi @ashwin-pc @kavilla , I have updated tests for this PR  although codecov report table is not regenerated and updated accordingly, please feel free to review. I will also try to add some cypress tests in functional test repo to make this more complete.

**Comment by raintygao** [comment]
Hi @ashwin-pc @kavilla , I have updated tests for this PR although codecov report table is not regenerated and updated accordingly, please feel free to review. 
I will also try to add some cypress tests in functional test repo to make this more complete.

**Comment by raintygao** [comment]
> [src/core/public/overlays/overlay_service.ts](https://app.codecov.io/gh/opensearch-project/OpenSearch-Dashboards/pull/5920?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opensearch-project#diff-c3JjL2NvcmUvcHVibGljL292ZXJsYXlzL292ZXJsYXlfc2VydmljZS50cw==)	0.00%	[4 Missing ⚠️ ](https://app.codecov.io/gh/opensearch-project/OpenSearch-Dashboards/pull/5920?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opensearch-project)
> [...re/public/overlays/sidecar/sidecar_service.mock.ts](https://app.codecov.io/gh/opensearch-project/OpenSearch-Dashboards/pull/5920?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opensearch-project#diff-c3JjL2NvcmUvcHVibGljL292ZXJsYXlzL3NpZGVjYXIvc2lkZWNhcl9zZXJ2aWNlLm1vY2sudHM=)	62.50%	[3 Missing ⚠️ ](https://app.codecov.io/gh/opensearch-project/OpenSearch-Dashboards/pull/5920?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=opensearch-project)

These two are originally missing and file to create mock for test.

**Comment by raintygao** [comment]
Hi @ashwin-pc @kavilla, most tests for this PR have been updated, please feel free to review.
I will also try to add some cypress tests in functional test repo to make this more complete.

File: src/core/public/overlays/sidecar/components/sidecar.tsx
Comment (SuZhou-Joe) [inline src/core/public/overlays/sidecar/components/sidecar.tsx:34]: Can we use `SIDECAR_DOCKED_MODE` here?

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (SuZhou-Joe) [inline src/core/public/overlays/sidecar/sidecar_service.tsx:91]: A little curious why these methods are not in SidecarRef, they seem like instance's methods.

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (SuZhou-Joe) [inline src/core/public/overlays/sidecar/sidecar_service.tsx:148]: This line will trigger the promise in line 156 I suppose, call `cleanupDom` one time in the promise should be enough?

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (SuZhou-Joe) [inline src/core/public/overlays/sidecar/sidecar_service.tsx:156]: Why don't we just listen to this.activeSidecar.onClose promise?

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (ruanyl) [inline src/core/public/overlays/sidecar/sidecar_service.tsx:11]:
```suggestion
import { BehaviorSubject, Observable, ReplaySubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
```

File: src/core/public/overlays/sidecar/components/resizable_button.tsx
Comment (wanglam) [inline src/core/public/overlays/sidecar/components/resizable_button.tsx:29]: How about add `initialFlyoutSize.current = flyoutSize` here. Then the `onMouseDown` doesn't need to depends on flyoutSize.

Comment (raintygao) [reply src/core/public/overlays/sidecar/components/resizable_button.tsx:29]: We couldn't, this will cause calculation error. And I think we should keep `initialFlyoutSize` assigning value only in `onMouseMove` to make it clear although there will be a more depence.

File: src/core/public/overlays/sidecar/sidecar_service.tsx
Comment (raintygao) [reply src/core/public/overlays/sidecar/sidecar_service.tsx:194]: This is mainly for class component. It manually call unmount for all components at target dom to avoid unmount not excuted as expected.

File: src/core/public/overlays/sidecar/components/sidecar.tsx
Comment (raintygao) [reply src/core/public/overlays/sidecar/components/sidecar.tsx:66]: I can understand you mean. I try to do some updates, but i find it's hard to decouple. 

`resizableButton` needs to get `dockedMode` to display vertical or horizontal style.  `onMouseMove` and `onMouseDown` both need to get `dockedMode` and `flyoutSize` to calculate new `paddingSize`, rather than only one event. So if we remove `dockedMode` and `flyoutSize`, we need to add other more parameters, like `isHorizontal`, `onMouseMove`,`onMouseDown`, i'm not sure if this will make it more confusing. 

I think there is a way to reduce parameters is to render buttons and containers in one component. I do some research on [similar component in other library](https://github.com/elastic/eui/blob/main/src-docs/src/views/resizable_container/resizable_button.tsx#L69), I find although it's named as `resizableButton`, it actually render both button and container.  If here we update to render container and button together, `resizableButton` will not be pure. So I think it could be difficult to update here.

File: src/core/public/overlays/sidecar/components/resizable_button.tsx
Comment (wanglam) [reply src/core/public/overlays/sidecar/components/resizable_button.tsx:29]: You're right. Need to add one more useRef to reduce efforts. The code may like
```ts
const flyoutSizeRef = useRef(flyoutSize);
flyoutSizeRef.current = flyoutSize;
const initialFlyoutSize = useRef(flyoutSize);

initialFlyoutSize.current = flyoutSizeRef.current;
```
Current implementation is good to me. It's a little bit complicated.

Comment (raintygao) [reply src/core/public/overlays/sidecar/components/resizable_button.tsx:29]: Yes, this way needs to add one more useRef.

**Comment by raintygao** [comment]
Hi Kevin @kgcreative, as I discussed with Ashwin above, there may exist some components in this PR that we can potentially support in OUI for code reuse in future and he would not block this PR because of this. I have promised him that I would try to follow up this in OUI when I have enough bandwidth in the future. So I think we could try to move forward this PR and release it as planned in 2.13, and try to follow up some updates in OUI in future, right? Do you think it's OK?

**Comment by raintygao** [comment]
Hi Kevin @kgcreative , as I discussed with Ashwin above, there may exist some components in this PR that we can potentially support in OUI for code reuse in future,  he would not block this PR because of this and I have promised him that I would try to follow up this in OUI when I have enough bandwidth in the future. So I think we could try to move forward this PR and release it as planned in 2.13, and try to follow up some updates in OUI in future. Do you have any concerns?

**Comment by raintygao** [comment]
Hi Kevin @kgcreative , as I discussed with Ashwin above, there may exist some components in this PR that we can potentially support in OUI for code reuse in future,  he would not block this PR because of this and I have promised him that I would try to follow up this in OUI when I have enough bandwidth in the future. So I think we could try to move forward this PR and release it as planned in 2.13, and try to follow up some updates in OUI in future. Do you have any concerns?

**Comment by raintygao** [comment]
Hi Kevin @kgcreative , as I discussed with Ashwin above, there may exist some components in this PR that we can potentially support in OUI for code reuse in future,  he replied that he would not block this PR because of this and I have promised him that I would try to follow up this in OUI when I have enough bandwidth in the future. 
So I think we could try to move forward this PR and release it as planned in 2.13, and try to follow up some updates in OUI in future. Do you have any concerns?

**Comment by kgcreative** [comment]
> Hi Kevin @kgcreative , as I discussed with Ashwin above, there may exist some components in this PR that we can potentially support in OUI for code reuse in future, he replied that he would not block this PR because of this and I have promised him that I would try to follow up this in OUI when I have enough bandwidth in the future. So I think we could try to move forward this PR and release it as planned in 2.13, and try to follow up some updates in OUI in future. Do you have any concerns?

As long as @ashwin-pc is on board, I'm supportive of this approach. This will unblock, and we can then make sure we can move forward with generalizing for OUI. I also agree that there may be some enhancements to flyout that can be made based on this work (For example, being able to resize flyout)

**Comment by ashwin-pc** [comment]
Yeah I'm okay with a follow-up task for OUI. Can you do a few things for this PR:

1. Can you add functional tests for the PR?
2. Can you add steps to the "Testing the changes" section of the PR?

This should help reviewing the PR faster.

**Comment by raintygao** [comment]
@ruanyl  I have created an issue in OUI to track this https://github.com/opensearch-project/oui/issues/1257. We could do component reuse. I can't be sure now that we can add a `rezableButton` component, at present it seems that `rezableFlyout` is more suitable, which is one of the reasons why I said that we need to follow up later, because we need to take some time to research on what components and how to design in OUI.

Github checks failed is not related to my PR, you could do a check on failed reasons. All flows are successful before I rebase main.

**Comment by raintygao** [comment]
> Yeah I'm okay with a follow-up task for OUI. Can you create the issue and link it here. Can you do a few things for this PR:
> 
> 1. Can you add functional tests for the PR?
> 2. Can you add steps to the "Testing the changes" section of the PR?
> 
> This should help reviewing the PR faster.

1.  Yes, working on this. PR will be created to FT repo.
2.  Updated in description.

**Comment by raintygao** [comment]
> Yeah I'm okay with a follow-up task for OUI. Can you create the issue and link it here. Can you do a few things for this PR:
> 
> 1. Can you add functional tests for the PR?
> 2. Can you add steps to the "Testing the changes" section of the PR?
> 
> This should help reviewing the PR faster.

@ashwin-pc 
1. Functional tests have been added.  https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/1141
2. Updated in description section of this PR.

File: src/core/public/overlays/sidecar/components/sidecar.tsx
Comment (raintygao) [reply src/core/public/overlays/sidecar/components/sidecar.tsx:7]: `yarn osd bootstrap` will fail because of this change. All import way of useObservable OSD is `import useObservable from 'react-use/lib/useObservable';`

**Comment by raintygao** [comment]
The only one failed check is fixed in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6119.

File: src/core/public/overlays/sidecar/components/sidecar.tsx
Comment (ruanyl) [reply src/core/public/overlays/sidecar/components/sidecar.tsx:7]: I see, thanks for the clarification, I didn't know that before 👍 Did a few investigation, it seems the types of `react-use` didn't work well with typescript.

**Review by SuZhou-Joe — DISMISSED** [review]
There are many UI changes happening in this PR within `core` directory, which should be moved to OUI, and I find the following issue has been raised: https://github.com/opensearch-project/oui/issues/1257 . It would be great that all the files under directory `src/core/public/overlays/sidecar/components` can be moved to OUI as standard components. Given the fact that @kgcreative wants a new sidecar mode, which can push the OSD instead of overlapping on the OSD content, to be available for all the plugins, current implementation is a trade-off for the tight 2.13 release.

I am giving the approval as I do not see blocker issues on the code change itself, but for the code design, I'd like to follow @ashwin-pc @kavilla 's opinion.

**Comment by raintygao** [comment]
> There are many UI changes happening in this PR within `core` directory, which should be moved to OUI, and I find the following issue has been raised: [opensearch-project/oui#1257](https://github.com/opensearch-project/oui/issues/1257) . It would be great that all the files under directory `src/core/public/overlays/sidecar/components` can be moved to OUI as standard components. Given the fact that @kgcreative wants a new sidecar mode, which can push the OSD instead of overlapping on the OSD content, to be available for all the plugins, current implementation is a trade-off for the tight 2.13 release.
> 
> I am giving the approval as I do not see blocker issues on the code change itself, but for the code design, I'd like to follow @ashwin-pc @kavilla 's opinion.

Thank you very much for your review and approval. For now I will ensure release first and follow up with component migration later.  I think this part needs some time to research and design, so I have created the issue https://github.com/opensearch-project/oui/issues/1257 and left some thoughts on it. I will share more thoughts and design on it with your guys about this part later.

**Review by ruanyl — DISMISSED** [review]
Approved, let's follow up future enhancement in https://github.com/opensearch-project/oui/issues/1257

**Comment by raintygao** [comment]
@SuZhou-Joe @ruanyl  I just rebased main to resolve conflicts in changelog, and your approvals were dismissed automatically. Could you approve again?

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Pull request merged**
